### PR TITLE
Updated the `LocalFileIO::ReplaceAlias` to substitute inplace

### PR DESCRIPTION
**Pull request by lumberyard-employee-dm**
Also added support to the JSONImporter to replace any aliases in the import path before appending it to the import path stack

This allows the `ResolveImports` function to be used with a path that
starts with a path that has the form of
`@gemroot:<gem-name>@/rest/of/path` and smartly update the import path
based on whether it the import filepath resolves to an absolute path in
which case it replaces the path importAbsPath value. Or if the import
path doesn't contain a fileio alias and is relative, it is appended to
the importAbsPath value.

Signed-off-by: lumberyard-employee-dm <56135373+lumberyard-employee-dm@users.noreply.github.com>

## How was this PR tested?

Added Unittest to validate inplace replacement of a FileIO alias at the beginning of a path

### CODE REVIEW

File: Code/Framework/AzFramework/AzFramework/IO/LocalFileIO.cpp [hunk 1]
```diff
@@ -829,8 +829,21 @@ namespace AZ
                     {
                         return false;
                     }
-                    replacedAliasPath.Native() = AZ::IO::FixedMaxPathString(AZStd::string_view{ aliasValue });
-                    replacedAliasPath.Native() += postAliasView;
+
+                    // Check if the path is input path does not overlap the output path
+                    if (postAliasView.data() < replacedAliasPath.Native().data()
+                        && postAliasView.data() >= replacedAliasPath.Native().data() + replacedAliasPath.Native().size())
```
Comment (AMZN-alexpete) [inline Code/Framework/AzFramework/AzFramework/IO/LocalFileIO.cpp:835]: I'm not sure I follow this logic.

If the `postAliasView.data()` pointer (unchanged part of path) is less than the `replacedAliasPath.Native().data()` (output path) pointer then how could that same pointer ever also be greater than the pointer to the output path + output path size?  Would something like this work?
```suggestion
                    if (postAliasView.data() + postAliasView.size() < replacedAliasPath.Native().data()
                        || postAliasView.data() >= replacedAliasPath.Native().data() + replacedAliasPath.Native().size())
```

Comment (lumberyard-employee-dm) [reply Code/Framework/AzFramework/AzFramework/IO/LocalFileIO.cpp:835]: Actually I inverted the check, but not the `&&` and `||`

Comment (AMZN-alexpete) [reply Code/Framework/AzFramework/AzFramework/IO/LocalFileIO.cpp]: ah this is breaking my brain.  It looks like `postAliasView` - which is a string view of `pathStrView` which comes from the input `path`, may be a view of a string completely unrelated to the output `replacedAliasPath`.   Like if you called the method this way:
```
                AZ::IO::FixedMaxPath pathWithAlias{"@engroot@/blah"};
                AZ::IO::FixedMaxPath pathResolved;
                AZ::IO::FileIOBase::GetInstance()->ReplaceAlias(pathResolved, AZ::IO::PathView{ pathWithAlias});
```

Is the intention that this method should only be called with a subview of the output path?

Comment (lumberyard-employee-dm) [reply Code/Framework/AzFramework/AzFramework/IO/LocalFileIO.cpp]: This function can be used with any input path.
The first block is when the input path isn't a subview of the output path.
The else block is when the input path is a subview of the output path.

**Review by AMZN-alexpete — APPROVED** [review]

File: Code/Framework/AzFramework/AzFramework/IO/LocalFileIO.cpp (hunk 1 above)
Comment (AMZN-alexpete) [reply Code/Framework/AzFramework/AzFramework/IO/LocalFileIO.cpp]: Thanks!  Makes a lot more sense now

**Review by AMZN-alexpete — APPROVED** [review]
Thank you!! This is huge for supporting cross-gem material $imports :)

File: Code/Framework/AzFramework/AzFramework/IO/LocalFileIO.cpp [hunk 2]
```diff
@@ -829,8 +829,21 @@ namespace AZ
                     {
                         return false;
                     }
-                    replacedAliasPath.Native() = AZ::IO::FixedMaxPathString(AZStd::string_view{ aliasValue });
-                    replacedAliasPath.Native() += postAliasView;
+
+                    // Check if the path is input path does not overlap the output path
```
Comment (amzn-tommy) [inline Code/Framework/AzFramework/AzFramework/IO/LocalFileIO.cpp:833]: nit: Should be "// Check if the input path does not overlap the output path

**Review by amzn-tommy — APPROVED** [review]

File: Code/Framework/AzFramework/AzFramework/IO/LocalFileIO.cpp (hunk 2 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzFramework/AzFramework/IO/LocalFileIO.cpp:833]: Will fix

**Pull request merged**